Conditional string templates don't work

Consider a simple StringTemplate example:

```java
public class TestConditionalStringTemplateLocator {

    @Rule
    public H2DatabaseRule db = new H2DatabaseRule().withPlugin(new SqlObjectPlugin());

    @Before
    public void setUp() throws Exception {
        Handle handle = db.getSharedHandle();
        handle.insert("insert into something (id, name) values (1, 'Martin')");
        handle.insert("insert into something (id, name) values (2, 'Joe')");
        handle.insert("insert into something (id, name) values (3, 'David')");
    }

    @Test
    public void testFindAndSortByName() {
        List<Integer> ids = db.getSharedHandle().attach(Dao.class).find(true, "name");
        System.out.println(ids);
    }

    @Test
    public void testFindWithoutSorting() {
        List<Integer> ids = db.getSharedHandle().attach(Dao.class).find(false, "");
        System.out.println(ids);
    }

    @UseStringTemplateSqlLocator
    interface Dao {
        @SqlQuery
        List<Integer> find(@Define("sort") boolean sort, @Define("sortBy") String sortBy);
    }
}
```

```
group Dao;

find(sort,sortBy) ::= << select id from something <if(sort)> order by <sortBy> <endif> >>
```

On the head of the `jdbi3` branch this test case returns the following exception:
```
org.jdbi.v3.core.exception.UnableToCreateStatementException: Exception parsing for named parameter replacement [statement:" select id from something <if(sortBy)> order by <sortBy> <endif> ", rewritten:"null", arguments:{ positional:{}, named:{}, finder:[]}]

	at org.jdbi.v3.core.rewriter.ColonPrefixStatementRewriter.rewrite(ColonPrefixStatementRewriter.java:62)
	at org.jdbi.v3.core.SqlStatement.internalExecute(SqlStatement.java:1236)
	at org.jdbi.v3.core.Query.iterator(Query.java:69)
	at org.jdbi.v3.core.ResultBearing.stream(ResultBearing.java:83)
	at org.jdbi.v3.core.ResultBearing.collect(ResultBearing.java:137)
	at org.jdbi.v3.sqlobject.ResultReturnThing$DefaultResultReturnThing.result(ResultReturnThing.java:118)
	at org.jdbi.v3.sqlobject.ResultReturnThing.map(ResultReturnThing.java:48)
	at org.jdbi.v3.sqlobject.QueryHandler.invoke(QueryHandler.java:42)
	at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$createMethodInterceptor$9(SqlObjectFactory.java:167)
	at org.jdbi.v3.stringtemplate.TestConditionalStringTemplateLocator$Dao$$EnhancerByCGLIB$$cd04f13c.find(<generated>)
	at org.jdbi.v3.stringtemplate.TestConditionalStringTemplateLocator.testFind(TestConditionalStringTemplateLocator.java:27)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:119)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: java.lang.IllegalArgumentException: MismatchedTokenException(40!=62)
	at org.jdbi.v3.core.internal.lexer.DefineStatementLexer.reportError(DefineStatementLexer.java:30)
	at org.antlr.runtime.Lexer.nextToken(Lexer.java:103)
	at org.jdbi.v3.core.rewriter.DefinedAttributeRewriter.rewriteDefines(DefinedAttributeRewriter.java:57)
	at org.jdbi.v3.core.rewriter.ColonPrefixStatementRewriter.rewrite(ColonPrefixStatementRewriter.java:59)
	... 39 more
Caused by: MismatchedTokenException(40!=62)
	at org.antlr.runtime.Lexer.match(Lexer.java:197)
	at org.jdbi.v3.core.internal.lexer.DefineStatementLexer.mGT(DefineStatementLexer.java:149)
	at org.jdbi.v3.core.internal.lexer.DefineStatementLexer.mDEFINE(DefineStatementLexer.java:543)
	at org.jdbi.v3.core.internal.lexer.DefineStatementLexer.mTokens(DefineStatementLexer.java:1116)
	at org.antlr.runtime.Lexer.nextToken(Lexer.java:89)
	... 41 more
```

This test case works with `jdbi3-3.0.0-alpha0`, so it looks like that the refactoring in #391 broke this functionality. 

At first glance, it looks like that we still need to pass a `StatementContext` to locators,
so `StringTemplateSqlLocator` can have access to attributes defined by the user and reevaluate string templates.

--- a/string-template/pom.xml
+++ b/string-template/pom.xml
@@ -79,6 +79,12 @@
         </dependency>
 
         <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
             <groupId>org.mockito</groupId>
             <artifactId>mockito-core</artifactId>
             <scope>test</scope>

--- a/string-template/src/main/java/org/jdbi/v3/stringtemplate/StringTemplateStatementRewriter.java
+++ b/string-template/src/main/java/org/jdbi/v3/stringtemplate/StringTemplateStatementRewriter.java
@@ -1,0 +1,57 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.stringtemplate;
+
+import static java.util.Objects.requireNonNull;
+
+import org.antlr.stringtemplate.StringTemplate;
+import org.antlr.stringtemplate.language.AngleBracketTemplateLexer;
+import org.jdbi.v3.core.Binding;
+import org.jdbi.v3.core.StatementContext;
+import org.jdbi.v3.core.rewriter.ColonPrefixStatementRewriter;
+import org.jdbi.v3.core.rewriter.RewrittenStatement;
+import org.jdbi.v3.core.rewriter.StatementRewriter;
+
+/**
+ * Rewrites a StringTemplate template, using the attributes on the {@link StatementContext} as template parameters.
+ */
+public class StringTemplateStatementRewriter implements StatementRewriter {
+    private final StatementRewriter delegate;
+
+    /**
+     * Constructs a rewriter which uses colon-prefixed parameter names.
+     */
+    public StringTemplateStatementRewriter() {
+        this(new ColonPrefixStatementRewriter());
+    }
+
+    /**
+     * Constructs a rewriter which uses the specified delegate rewriter, after parsing out any StringTemplate
+     * expressions.
+     *
+     * @param delegate the delegate rewriter.
+     */
+    public StringTemplateStatementRewriter(StatementRewriter delegate) {
+        this.delegate = requireNonNull(delegate);
+    }
+
+    @Override
+    public RewrittenStatement rewrite(String sql, Binding params, StatementContext ctx) {
+        StringTemplate template = new StringTemplate(sql, AngleBracketTemplateLexer.class);
+
+        ctx.getAttributes().forEach(template::setAttribute);
+
+        return delegate.rewrite(template.toString(), params, ctx);
+    }
+}

--- a/string-template/src/main/java/org/jdbi/v3/stringtemplate/UseStringTemplateSqlLocator.java
+++ b/string-template/src/main/java/org/jdbi/v3/stringtemplate/UseStringTemplateSqlLocator.java
@@ -18,13 +18,19 @@ import java.lang.annotation.ElementType;
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
+import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.util.function.Consumer;
 
+import org.jdbi.v3.core.rewriter.ColonPrefixStatementRewriter;
+import org.jdbi.v3.core.rewriter.StatementRewriter;
 import org.jdbi.v3.sqlobject.SqlAnnotations;
 import org.jdbi.v3.sqlobject.SqlObjectConfig;
 import org.jdbi.v3.sqlobject.SqlObjectConfigurerFactory;
 import org.jdbi.v3.sqlobject.SqlObjectConfiguringAnnotation;
+import org.jdbi.v3.sqlobject.SqlStatementCustomizer;
+import org.jdbi.v3.sqlobject.SqlStatementCustomizerFactory;
+import org.jdbi.v3.sqlobject.SqlStatementCustomizingAnnotation;
 import org.jdbi.v3.sqlobject.locator.SqlLocator;
 
 /**
@@ -44,11 +50,14 @@ import org.jdbi.v3.sqlobject.locator.SqlLocator;
  *     }
  * </pre>
  */
-@SqlObjectConfiguringAnnotation(UseStringTemplateSqlLocator.Factory.class)
+@SqlObjectConfiguringAnnotation(UseStringTemplateSqlLocator.LocatorFactory.class)
+@SqlStatementCustomizingAnnotation(UseStringTemplateSqlLocator.RewriterFactory.class)
 @Retention(RetentionPolicy.RUNTIME)
-@Target({ElementType.TYPE})
+@Target({ElementType.TYPE, ElementType.METHOD})
 public @interface UseStringTemplateSqlLocator {
-    class Factory implements SqlObjectConfigurerFactory {
+    Class<? extends StatementRewriter> value() default ColonPrefixStatementRewriter.class;
+
+    class LocatorFactory implements SqlObjectConfigurerFactory {
         private static final SqlLocator SQL_LOCATOR = (sqlObjectType, method) -> {
             String name = SqlAnnotations.getAnnotationValue(method).orElseGet(method::getName);
             return StringTemplateSqlLocator.findStringTemplateSql(sqlObjectType, name);
@@ -64,6 +73,32 @@ public @interface UseStringTemplateSqlLocator {
         @Override
         public Consumer<SqlObjectConfig> createForMethod(Annotation annotation, Class<?> sqlObjectType, Method method) {
             return CONFIGURER;
+        }
+    }
+
+    class RewriterFactory implements SqlStatementCustomizerFactory {
+        @Override
+        public SqlStatementCustomizer createForType(Annotation annotation, Class<?> sqlObjectType) {
+            return create((UseStringTemplateSqlLocator) annotation);
+        }
+
+        @Override
+        public SqlStatementCustomizer createForMethod(Annotation annotation, Class<?> sqlObjectType, Method method) {
+            return create((UseStringTemplateSqlLocator) annotation);
+        }
+
+        private SqlStatementCustomizer create(UseStringTemplateSqlLocator annotation) {
+            StatementRewriter delegate = createDelegate(annotation.value());
+            StatementRewriter rewriter = new StringTemplateStatementRewriter(delegate);
+            return q -> q.setStatementRewriter(rewriter);
+        }
+
+        private StatementRewriter createDelegate(Class<? extends StatementRewriter> type) {
+            try {
+                return type.getConstructor().newInstance();
+            } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+                throw new IllegalStateException("Error instantiating delegate statement rewriter: " + type, e);
+            }
         }
     }
 }

--- a/string-template/src/main/java/org/jdbi/v3/stringtemplate/UseStringTemplateStatementRewriter.java
+++ b/string-template/src/main/java/org/jdbi/v3/stringtemplate/UseStringTemplateStatementRewriter.java
@@ -1,0 +1,66 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.stringtemplate;
+
+import java.lang.annotation.Annotation;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import org.jdbi.v3.core.rewriter.ColonPrefixStatementRewriter;
+import org.jdbi.v3.core.rewriter.StatementRewriter;
+import org.jdbi.v3.sqlobject.SqlStatementCustomizer;
+import org.jdbi.v3.sqlobject.SqlStatementCustomizerFactory;
+import org.jdbi.v3.sqlobject.SqlStatementCustomizingAnnotation;
+
+/**
+ * Configures a SQL object class or method to rewrite SQL statements using StringTemplate. Method parameters annotated
+ * with {@link org.jdbi.v3.sqlobject.customizers.Define @Define} are passed to the StringTemplate as template
+ * attributes.
+ */
+@SqlStatementCustomizingAnnotation(UseStringTemplateStatementRewriter.Factory.class)
+@Target({ElementType.TYPE, ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface UseStringTemplateStatementRewriter {
+    Class<? extends StatementRewriter> value() default ColonPrefixStatementRewriter.class;
+
+    class Factory implements SqlStatementCustomizerFactory {
+        @Override
+        public SqlStatementCustomizer createForType(Annotation annotation, Class<?> sqlObjectType) {
+            return create((UseStringTemplateStatementRewriter) annotation);
+        }
+
+        @Override
+        public SqlStatementCustomizer createForMethod(Annotation annotation, Class<?> sqlObjectType, Method method) {
+            return create((UseStringTemplateStatementRewriter) annotation);
+        }
+
+        private SqlStatementCustomizer create(UseStringTemplateStatementRewriter annotation) {
+            StatementRewriter delegate = createDelegate(annotation.value());
+            StatementRewriter rewriter = new StringTemplateStatementRewriter(delegate);
+            return q -> q.setStatementRewriter(rewriter);
+        }
+
+        private StatementRewriter createDelegate(Class<? extends StatementRewriter> type) {
+            try {
+                return type.getConstructor().newInstance();
+            } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+                throw new IllegalStateException("Error instantiating delegate statement rewriter: " + type, e);
+            }
+        }
+    }
+}

--- a/string-template/src/test/java/org/jdbi/v3/stringtemplate/TestConditionalStringTemplateLocator.java
+++ b/string-template/src/test/java/org/jdbi/v3/stringtemplate/TestConditionalStringTemplateLocator.java
@@ -1,0 +1,75 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.stringtemplate;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+
+import org.jdbi.v3.core.H2DatabaseRule;
+import org.jdbi.v3.core.Handle;
+import org.jdbi.v3.sqlobject.SqlObjectPlugin;
+import org.jdbi.v3.sqlobject.SqlQuery;
+import org.jdbi.v3.sqlobject.customizers.Define;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class TestConditionalStringTemplateLocator {
+
+    @Rule
+    public H2DatabaseRule db = new H2DatabaseRule().withPlugin(new SqlObjectPlugin());
+
+    @Before
+    public void setUp() throws Exception {
+        Handle handle = db.getSharedHandle();
+        handle.insert("insert into something (id, name) values (1, 'Martin')");
+        handle.insert("insert into something (id, name) values (3, 'David')");
+        handle.insert("insert into something (id, name) values (2, 'Joe')");
+    }
+
+    @Test
+    public void testLocateFindAndSortByName() {
+        List<Integer> ids = db.getSharedHandle().attach(Dao.class).findLocated(true, "name");
+        assertThat(ids).containsExactly(3, 2, 1);
+    }
+
+    @Test
+    public void testLocateFindWithoutSorting() {
+        List<Integer> ids = db.getSharedHandle().attach(Dao.class).findLocated(false, "");
+        assertThat(ids).containsExactly(1, 2, 3);
+    }
+
+    @Test
+    public void testInlineFindAndSortByName() {
+        List<Integer> ids = db.getSharedHandle().attach(Dao.class).findInline(true, "name");
+        assertThat(ids).containsExactly(3, 2, 1);
+    }
+
+    @Test
+    public void testInlineFindWithoutSorting() {
+        List<Integer> ids = db.getSharedHandle().attach(Dao.class).findInline(false, "");
+        assertThat(ids).containsExactly(1, 2, 3);
+    }
+
+    interface Dao {
+        @SqlQuery
+        @UseStringTemplateSqlLocator
+        List<Integer> findLocated(@Define("sort") boolean sort, @Define("sortBy") String sortBy);
+
+        @SqlQuery("select id from something order by <if(sort)> <sortBy>, <endif> id")
+        @UseStringTemplateStatementRewriter
+        List<Integer> findInline(@Define("sort") boolean sort, @Define("sortBy") String sortBy);
+    }
+}

--- a/string-template/src/test/resources/org/jdbi/v3/stringtemplate/TestConditionalStringTemplateLocator$Dao.sql.stg
+++ b/string-template/src/test/resources/org/jdbi/v3/stringtemplate/TestConditionalStringTemplateLocator$Dao.sql.stg
@@ -1,0 +1,3 @@
+group Dao;
+
+findLocated(sort,sortBy) ::= << select id from something order by <if(sort)> <sortBy>, <endif> id >>

